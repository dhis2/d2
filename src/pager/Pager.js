import { isDefined } from '../lib/check'

/**
 * @description
 * Pager object that can be used to navigate pages within a `Modelcollection`
 *
 * @memberof module:model
 */

/**
 * @memberof module:model
 */
// TODO: Move to the model map
class Pager {
    /**
     * @constructor
     *
     * @param {Object} [pager={page: 1, pageCount: 1}] Paging information object.
     * @param {Object} [pagingHandler={list: () => Promise.reject('No handler available')}] Paging handler object. The requirement for this object is that it has a list method.
     *
     * @description
     * Returns a newly created pager object with methods to navigate pages.
     */
    constructor(
        pager = { page: 1, pageCount: 1, query: {} },
        pagingHandler = { list: () => Promise.reject('No handler available') }
    ) {
        /**
         * @property {number} page Current page number
         */
        this.page = pager.page

        /**
         * @property {number} pageCount The total number of pages available
         */
        this.pageCount = pager.pageCount

        /**
         * @property {number} total The total number of items available.
         *
         * @description
         * This represents the total number of items available in the system. Note it is not the number of items
         * on the current page.
         */
        this.total = pager.total

        /**
         * @property {string} nextPage The url to the next page.
         *
         * @description
         * If there is no next page then this will be undefined.
         */
        this.nextPage = pager.nextPage

        /**
         * @property {string} prevPage The url to the previous page
         *
         * @description
         * If there is no previous page then this will be undefined.
         */
        this.prevPage = pager.prevPage

        /**
         * @property {object} query Query parameters
         *
         * @description
         * Query parameters are used for things like filtering and field selection. Used to guarantee that pages are
         * from the same collection.
         */
        this.query = pager.query

        this.pagingHandler = pagingHandler
    }

    /**
     * @returns {Boolean} Result is true when there is a next page, false when there is not.
     *
     * @description
     * Check whether there is a next page.
     */
    hasNextPage() {
        return isDefined(this.nextPage)
    }

    /**
     * Check whether there is a previous page.
     *
     * @returns {Boolean} Result is true when there is a previous page, false when there is not.
     */
    hasPreviousPage() {
        return isDefined(this.prevPage)
    }

    /**
     * @description
     * Loads the next page in the collection if there is one. If no additional pages are available the Promise will reject.
     *
     * @returns {Promise} Promise that resolves with a new `ModelCollection` containing the next page's data. Or rejects with
     * a string when there is no next page for this collection or when the request for the next page failed.
     *
     * @example
     * d2.models.organisationUnit
     *   .list()
     *   .then(collection => {
     *     collection.pager.getNextPage()
     *       .then(secondPageCollection => {
     *         console.log(secondPageCollection.toArray());
     *       });
     *   });
     */
    getNextPage() {
        if (this.hasNextPage()) {
            return this.goToPage(this.page + 1)
        }
        return Promise.reject('There is no next page for this collection')
    }

    /**
     * @description
     * Loads the previous page in the collection if there is one. If no previous pages are available the Promise will reject.
     *
     * @returns {Promise} Promise that resolves with a new `ModelCollection` containing the previous page's data. Or rejects with
     * a string when there is no previous page for this collection or when the request for the previous page failed.
     *
     * @example
     * d2.models.organisationUnit
     *   .list()
     *   .then(collection => {
     *     collection.pager.goToPage(3)
     *       .then(collection => collection.pager.getPreviousPage())
     *       .then(secondPageCollection => {
     *         console.log(secondPageCollection.toArray());
     *       });
     *   });
     */
    getPreviousPage() {
        if (this.hasPreviousPage()) {
            return this.goToPage(this.page - 1)
        }
        return Promise.reject('There is no previous page for this collection')
    }

    /**
     * Loads a specific page for the collection. If the requested page is out of the range of available pages (e.g < 0 or > page count)
     * the Promise will reject with an error.
     *
     * @param {Number} pageNr The number of the page you wish to navigate to.
     * @returns {Promise} Promise that resolves with a new `ModelCollection` containing the data for the requested page.
     *
     * @example
     * d2.models.organisationUnit
     *   .list()
     *   .then(collection => {
     *     collection.pager.goToPage(4)
     *       .then(fourthPageCollection => {
     *         console.log(fourthPageCollection.toArray());
     *       });
     *   });
     */
    // TODO: Throwing the errors here is not really consistent with the rejection of promises for the getNextPage and getPreviousPage
    goToPage(pageNr) {
        if (pageNr < 1) {
            throw new Error('PageNr can not be less than 1')
        }
        if (pageNr > this.pageCount) {
            throw new Error(
                `PageNr can not be larger than the total page count of ${
                    this.pageCount
                }`
            )
        }

        return this.pagingHandler.list(
            Object.assign({}, this.query, { page: pageNr })
        )
    }
}

export default Pager
