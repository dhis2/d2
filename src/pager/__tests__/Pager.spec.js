import Pager from '../Pager'

describe('Pager', () => {
    let pagerFixtureOne
    let pageFixtureTwo

    it('should not be allowed to be called without new', () => {
        expect(() => Pager()).toThrowErrorMatchingSnapshot()
    })

    describe('instance without data', () => {
        let pager

        beforeEach(() => {
            pager = new Pager()
        })

        it('should set the page to first', () => {
            expect(pager.page).toBe(1)
        })

        it('should set the total page count to 1', () => {
            expect(pager.pageCount).toBe(1)
        })

        it('should set the total item count to undefined', () => {
            expect(pager.total).toBeUndefined()
        })

        it('should not set the nextPage', () => {
            expect(pager.nextPage).toBeUndefined()
        })

        it('should not set the prevPage', () => {
            expect(pager.prevPage).toBeUndefined()
        })
    })

    describe('instance with data', () => {
        let pager
        let modelDefinition

        beforeEach(() => {
            pagerFixtureOne = {
                page: 1,
                pageCount: 37,
                total: 1844,
                nextPage: 'http://localhost:8080/dhis/api/dataElements?page=2',
            }
            pageFixtureTwo = {
                page: 3,
                pageCount: 37,
                total: 1844,
                nextPage: 'http://localhost:8080/dhis/api/dataElements?page=4',
                prevPage: 'http://localhost:8080/dhis/api/dataElements?page=2',
            }

            class ModelDefinition {}
            ModelDefinition.prototype.list = jest
                .fn()
                .mockReturnValue(Promise.resolve())
            modelDefinition = new ModelDefinition()

            pager = new Pager(pagerFixtureOne, modelDefinition)
        })

        it('should be an instance of Pager', () => {
            expect(pager).toBeInstanceOf(Pager)
        })

        it('should have a total item count', () => {
            expect(pager.total).toBe(1844)
        })

        it('should have the current page number', () => {
            expect(pager.page).toBe(1)
        })

        it('should have a pageCount', () => {
            expect(pager.pageCount).toBe(37)
        })

        it('should have a nextPage url', () => {
            expect(pager.nextPage).toBe(
                'http://localhost:8080/dhis/api/dataElements?page=2'
            )
        })

        it('should have previous page', () => {
            pager = new Pager(pageFixtureTwo)

            expect(pager.prevPage).toBe(
                'http://localhost:8080/dhis/api/dataElements?page=2'
            )
        })

        describe('hasNextPage', () => {
            it('should be a function', () => {
                expect(pager.hasNextPage).toBeInstanceOf(Function)
            })

            it('should return true if there is a next page', () => {
                expect(pager.hasNextPage()).toBe(true)
            })

            it('should return false if there is no next page', () => {
                delete pager.nextPage

                expect(pager.hasNextPage()).toBe(false)
            })
        })

        describe('hasPreviousPage', () => {
            it('should be a function', () => {
                expect(pager.hasPreviousPage).toBeInstanceOf(Function)
            })

            it('should return true if there is a previous page', () => {
                pager.prevPage = 'some link to a page'

                expect(pager.hasPreviousPage()).toBe(true)
            })

            it('should return false if there is no previous page', () => {
                expect(pager.hasPreviousPage()).toBe(false)
            })
        })

        describe('nextPage', () => {
            it('should be a method on the collection', () => {
                expect(pager.getNextPage).toBeInstanceOf(Function)
            })

            it('should return a promise', () => {
                expect(pager.getNextPage()).toBeInstanceOf(Promise)
            })

            it('should call the model definition for a new list', () => {
                pager.getNextPage()

                expect(modelDefinition.list).toBeCalled()
            })

            it('should only ask for a new list if the pager has a nextPage property', () => {
                delete pager.nextPage

                return pager.getNextPage().catch(() => {
                    expect(modelDefinition.list).not.toHaveBeenCalled()
                })
            })

            it('should return a rejected promise if there are no more new pages', () => {
                delete pager.nextPage

                return pager.getNextPage().catch(message => {
                    expect(message).toBe(
                        'There is no next page for this collection'
                    )
                })
            })

            it('should call next page with the current page number + 1', () => {
                pager.getNextPage()

                expect(modelDefinition.list).toBeCalledWith({ page: 2 })
            })
        })

        describe('previousPage', () => {
            it('should be a method on the collection', () => {
                expect(pager.getPreviousPage).toBeInstanceOf(Function)
            })

            it('should ask for the previous page if the prevPage property is set', () => {
                pager.page = 2
                pager.prevPage = 'http://url.to.the.next.page'

                expect.assertions(1)

                return pager.getPreviousPage().then(() => {
                    expect(modelDefinition.list).toBeCalled()
                })
            })

            it('should not ask for a new list if there is no previous page', () => {
                expect(modelDefinition.list).not.toHaveBeenCalled()
            })

            it('should return a rejected promise if there are no more previous pages', () => {
                expect.assertions(1)

                return pager.getPreviousPage().catch(message => {
                    expect(message).toBe(
                        'There is no previous page for this collection'
                    )
                })
            })

            it('should call the list method with the current page number - 1', () => {
                pager.page = 3
                pager.prevPage = 'http://url.to.the.next.page'

                expect.assertions(1)

                return pager.getPreviousPage().then(() => {
                    expect(modelDefinition.list).toBeCalledWith({ page: 2 })
                })
            })
        })

        describe('goToPage', () => {
            it('should call the list method with the passed page number', () => {
                pager.goToPage(2)

                expect(modelDefinition.list).toBeCalledWith({ page: 2 })
            })

            it('should throw an error when the page is less than 1', () => {
                expect(() => pager.goToPage(0)).toThrowError(
                    'PageNr can not be less than 1'
                )
                expect(() => pager.goToPage(-1)).toThrowError(
                    'PageNr can not be less than 1'
                )
            })

            it('should throw an error when the page is larger than the pagecount', () => {
                expect(() => pager.goToPage(38)).toThrowError(
                    'PageNr can not be larger than the total page count of 37'
                )
                expect(() => pager.goToPage(100)).toThrowError(
                    'PageNr can not be larger than the total page count of 37'
                )
            })
        })

        describe('should throw error when there is no page handler', () => {
            it('should throw an error when no handler is specified', () => {
                pager = new Pager(pagerFixtureOne)

                expect.assertions(1)

                return pager.getNextPage().catch(() => {
                    // TODO: useless assertion
                    expect(true).toBe(true)
                })
            })
        })
    })

    describe('instance with data and query parameters', () => {
        let pager
        let modelDefinition

        beforeEach(() => {
            pagerFixtureOne = {
                page: 1,
                pageCount: 37,
                query: { fields: ':all' },
                total: 1844,
                nextPage: 'http://localhost:8080/dhis/api/dataElements?page=2l',
            }
            pageFixtureTwo = {
                page: 3,
                pageCount: 37,
                total: 1844,
                nextPage: 'http://localhost:8080/dhis/api/dataElements?page=4',
                prevPage: 'http://localhost:8080/dhis/api/dataElements?page=2',
            }

            class ModelDefinition {}
            ModelDefinition.prototype.list = jest
                .fn()
                .mockReturnValue(Promise.resolve())
            modelDefinition = new ModelDefinition()

            pager = new Pager(pagerFixtureOne, modelDefinition, {
                fields: ':all',
            })
        })

        describe('nextPage', () => {
            it('should include the current query parameters', () => {
                pager.getNextPage()

                expect(modelDefinition.list).toBeCalledWith({
                    page: 2,
                    fields: ':all',
                })
            })
        })

        describe('previousPage', () => {
            it('should include the current query parameters', () => {
                pager.page = 3
                pager.prevPage = 'http://url.to.the.next.page'

                pager.getPreviousPage()

                expect(modelDefinition.list).toBeCalledWith({
                    page: 2,
                    fields: ':all',
                })
            })
        })

        describe('goToPage', () => {
            it('should call the list method with the current query parameters', () => {
                pager.goToPage(2)

                expect(modelDefinition.list).toBeCalledWith({
                    page: 2,
                    fields: ':all',
                })
            })
        })
    })
})
