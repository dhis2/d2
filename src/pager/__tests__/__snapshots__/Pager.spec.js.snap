// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager should not be allowed to be called without new 1`] = `"Class constructor Pager cannot be invoked without 'new'"`;
